### PR TITLE
WFLY-15088 With transactional cache, expiration scheduling reads out of date maxInactiveInterval value changed on non-primary owner

### PR DESCRIPTION
**Pull request by pferraro**
https://issues.redhat.com/browse/WFLY-15088

### CODE REVIEW

**Comment by pferraro** [comment]
@rhusar Please review.

**Review by rhusar — CHANGES_REQUESTED** [review]
Needs changes - looks like the marshaller registration is missing - the scheduling fails with the following:

```
2021-08-16 01:49:16,880 INFO  [org.wildfly.clustering.ee.infinispan] (default task-2) WFLYCLEEINF0002: Failed to schedule jdktSVxbm6e1sCZ06bJcSmPQRwmDsGnFX6dHsc_H on primary owner.: java.lang.IllegalArgumentException: No marshaller registered for Java type org.wildfly.clustering.web.infinispan.session.SimpleSessionExpirationMetaData
	at org.infinispan.protostream.impl.SerializationContextImpl.getMarshallerDelegate(SerializationContextImpl.java:483)
	at org.infinispan.protostream.impl.SerializationContextImpl.getMarshaller(SerializationContextImpl.java:446)
	at org.wildfly.clustering.marshalling.protostream.ProtoStreamWriter.findMarshaller(ProtoStreamWriter.java:99)
	at org.wildfly.clustering.marshalling.protostream.AnyMarshaller.getField(AnyMarshaller.java:112)
	at org.wildfly.clustering.marshalling.protostream.AnyMarshaller.writeTo(AnyMarshaller.java:74)
	at org.wildfly.clustering.marshalling.protostream.AnyMarshaller.writeTo(AnyMarshaller.java:36)
	at org.wildfly.clustering.marshalling.protostream.Marshallable.size(Marshallable.java:62)
	at org.wildfly.clustering.marshalling.protostream.SizeComputingProtoStreamWriter.writeObjectNoTag(SizeComputingProtoStreamWriter.java:80)
	at org.wildfly.clustering.marshalling.protostream.ProtoStreamWriter.writeObject(ProtoStreamWriter.java:62)
	at org.wildfly.clustering.marshalling.protostream.ProtoStreamWriter.writeAny(ProtoStreamWriter.java:50)
	at org.wildfly.clustering.ee.infinispan.scheduler.ScheduleWithMetaDataCommandMarshaller.writeTo(ScheduleWithMetaDataCommandMarshaller.java:76)
	at org.wildfly.clustering.ee.infinispan.scheduler.ScheduleWithMetaDataCommandMarshaller.writeTo(ScheduleWithMetaDataCommandMarshaller.java:36)
	at org.wildfly.clustering.marshalling.protostream.Marshallable.size(Marshallable.java:62)
	at org.wildfly.clustering.marshalling.protostream.SizeComputingProtoStreamWriter.writeObjectNoTag(SizeComputingProtoStreamWriter.java:80)
	at org.wildfly.clustering.marshalling.protostream.ProtoStreamWriter.writeObject(ProtoStreamWriter.java:62)
	at org.wildfly.clustering.marshalling.protostream.TypedObjectMarshaller.writeTo(TypedObjectMarshaller.java:60)
	at org.wildfly.clustering.marshalling.protostream.AnyMarshaller.writeTo(AnyMarshaller.java:76)
	at org.wildfly.clustering.marshalling.protostream.AnyMarshaller.writeTo(AnyMarshaller.java:36)
	at org.wildfly.clustering.marshalling.protostream.Marshallable.size(Marshallable.java:62)
	at org.wildfly.clustering.marshalling.protostream.SizeComputingProtoStreamWriter.writeObjectNoTag(SizeComputingProtoStreamWriter.java:80)
	at org.wildfly.clustering.marshalling.protostream.ProtoStreamByteBufferMarshaller.size(ProtoStreamByteBufferMarshaller.java:51)
	at org.wildfly.clustering.marshalling.spi.ByteBufferMarshaller.write(ByteBufferMarshaller.java:53)
	at org.wildfly.clustering.marshalling.spi.ByteBufferMarshalledValue.getBuffer(ByteBufferMarshalledValue.java:68)
	at org.wildfly.clustering.marshalling.protostream.FunctionalScalarMarshaller.writeTo(FunctionalScalarMarshaller.java:169)
	at org.wildfly.clustering.marshalling.protostream.Marshallable.size(Marshallable.java:62)
	at org.wildfly.clustering.marshalling.protostream.SizeComputingProtoStreamWriter.writeObjectNoTag(SizeComputingProtoStreamWriter.java:80)
	at org.wildfly.clustering.marshalling.protostream.ProtoStreamWriter.writeObject(ProtoStreamWriter.java:62)
	at org.wildfly.clustering.marshalling.protostream.TypedObjectMarshaller.writeTo(TypedObjectMarshaller.java:60)
	at org.wildfly.clustering.marshalling.protostream.AnyMarshaller.writeTo(AnyMarshaller.java:76)
	at org.wildfly.clustering.marshalling.protostream.AnyMarshaller.writeTo(AnyMarshaller.java:36)
	at org.wildfly.clustering.marshalling.protostream.Marshallable.size(Marshallable.java:62)
	at org.wildfly.clustering.marshalling.protostream.SizeComputingProtoStreamWriter.writeObjectNoTag(SizeComputingProtoStreamWriter.java:80)
	at org.wildfly.clustering.marshalling.protostream.ProtoStreamWriter.writeObject(ProtoStreamWriter.java:62)
	at org.wildfly.clustering.marshalling.protostream.ProtoStreamWriter.writeAny(ProtoStreamWriter.java:50)
	at org.wildfly.clustering.marshalling.protostream.util.MapEntryMarshaller.writeTo(MapEntryMarshaller.java:84)
	at org.wildfly.clustering.marshalling.protostream.util.MapEntryMarshaller.writeTo(MapEntryMarshaller.java:40)
	at org.wildfly.clustering.marshalling.protostream.Marshallable.size(Marshallable.java:62)
	at org.wildfly.clustering.marshalling.protostream.SizeComputingProtoStreamWriter.writeObjectNoTag(SizeComputingProtoStreamWriter.java:80)
	at org.wildfly.clustering.marshalling.protostream.ProtoStreamWriter.writeObject(ProtoStreamWriter.java:62)
	at org.wildfly.clustering.marshalling.protostream.TypedObjectMarshaller.writeTo(TypedObjectMarshaller.java:60)
	at org.wildfly.clustering.marshalling.protostream.AnyMarshaller.writeTo(AnyMarshaller.java:76)
	at org.wildfly.clustering.marshalling.protostream.AnyMarshaller.writeTo(AnyMarshaller.java:36)
	at org.wildfly.clustering.marshalling.protostream.Marshallable.size(Marshallable.java:62)
	at org.wildfly.clustering.marshalling.protostream.SizeComputingProtoStreamWriter.writeObjectNoTag(SizeComputingProtoStreamWriter.java:80)
	at org.wildfly.clustering.marshalling.protostream.ProtoStreamByteBufferMarshaller.size(ProtoStreamByteBufferMarshaller.java:51)
	at org.wildfly.clustering.marshalling.spi.ByteBufferMarshaller.write(ByteBufferMarshaller.java:53)
	at org.wildfly.clustering.server.dispatcher.CommandDispatcherMarshaller.marshal(CommandDispatcherMarshaller.java:54)
	at org.wildfly.clustering.server.dispatcher.ChannelCommandDispatcher.createBuffer(ChannelCommandDispatcher.java:145)
	at org.wildfly.clustering.server.dispatcher.ChannelCommandDispatcher.executeOnMember(ChannelCommandDispatcher.java:109)
	at org.wildfly.clustering.server.dispatcher.ManagedCommandDispatcher.executeOnMember(ManagedCommandDispatcher.java:53)
	at org.wildfly.clustering.ee.infinispan.scheduler.PrimaryOwnerScheduler$1.get(PrimaryOwnerScheduler.java:90)
	at org.wildfly.clustering.ee.infinispan.scheduler.PrimaryOwnerScheduler$1.get(PrimaryOwnerScheduler.java:85)
	at org.wildfly.clustering.ee.cache.retry.RetryingInvoker.invoke(RetryingInvoker.java:72)
	at org.wildfly.clustering.ee.infinispan.scheduler.PrimaryOwnerScheduler.executeOnPrimaryOwner(PrimaryOwnerScheduler.java:93)
	at org.wildfly.clustering.ee.infinispan.scheduler.PrimaryOwnerScheduler.schedule(PrimaryOwnerScheduler.java:66)
	at org.wildfly.clustering.web.infinispan.session.InfinispanSessionManager$1.accept(InfinispanSessionManager.java:111)
	at org.wildfly.clustering.web.infinispan.session.InfinispanSessionManager$1.accept(InfinispanSessionManager.java:107)
	at org.wildfly.clustering.web.cache.session.ValidSession.close(ValidSession.java:90)
	at org.wildfly.clustering.web.cache.session.ConcurrentSessionManager$ConcurrentSession.closeSession(ConcurrentSessionManager.java:159)
	at org.wildfly.clustering.web.cache.session.ConcurrentSessionManager$1.accept(ConcurrentSessionManager.java:55)
	at org.wildfly.clustering.web.cache.session.ConcurrentSessionManager$1.accept(ConcurrentSessionManager.java:52)
	at org.wildfly.clustering.ee.cache.SimpleManager$1.run(SimpleManager.java:55)
	at org.wildfly.clustering.web.cache.session.ConcurrentSessionManager$ConcurrentSession.close(ConcurrentSessionManager.java:190)
	at org.wildfly.clustering.web.undertow.session.DistributableSession.requestDone(DistributableSession.java:120)
	at io.undertow.servlet.spec.ServletContextImpl.updateSessionAccessTime(ServletContextImpl.java:987)
	at io.undertow.servlet.spec.HttpServletResponseImpl.responseDone(HttpServletResponseImpl.java:590)
	at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:339)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:79)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:134)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:131)
	at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
	at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
	at org.wildfly.extension.undertow.security.SecurityContextThreadSetupAction.lambda$create$0(SecurityContextThreadSetupAction.java:105)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1524)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1524)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1524)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1524)
	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:260)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$000(ServletInitialHandler.java:79)
	at io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:100)
	at io.undertow.server.Connectors.executeRootHandler(Connectors.java:387)
	at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:852)
	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1990)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1486)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1377)
	at org.xnio.XnioWorker$WorkerThreadFactory$1$1.run(XnioWorker.java:1280)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by rhusar** [comment]
BTW Also demonstrates why we need those changes in this test in https://github.com/wildfly/wildfly/pull/14533 since it passes on the PR since the test doesn't actually test scheduling expiration.

File: clustering/web/infinispan/src/main/java/org/wildfly/clustering/web/infinispan/session/SessionSerializationContextInitializer.java [hunk 1]
```diff
@@ -40,5 +40,6 @@ public void registerMarshallers(SerializationContext context) {
         context.registerMarshaller(new SessionKeyMarshaller<>(SessionCreationMetaDataKey.class, SessionCreationMetaDataKey::new));
         context.registerMarshaller(new SessionKeyMarshaller<>(SessionAccessMetaDataKey.class, SessionAccessMetaDataKey::new));
         context.registerMarshaller(new EnumMarshaller<>(SessionCreationMetaDataKeyFilter.class));
+        context.registerMarshaller(new SimpleSessionExpirationMetaDataMarshaller());
```
[the remark this answers is not in the document]
Comment (pferraro) [reply clustering/web/infinispan/src/main/java/org/wildfly/clustering/web/infinispan/session/SessionSerializationContextInitializer.java:43]: @rhusar Fixed.

[the remark this answers is not in the document]
Comment (rhusar) [reply clustering/web/infinispan/src/main/java/org/wildfly/clustering/web/infinispan/session/SessionSerializationContextInitializer.java:43]: @pferraro Please push :-)

**Comment by rhusar** [comment]
@pferraro Hm, the latest linux/ely/jdk8 run failed with what we have been seeing before - might be related.

```
&amp#27;[0m&amp#27;[31m08:12:57,039 ERROR [org.jboss.as.controller.management-operation] (management-handler-thread - 1) WFLYCTL0348: Timeout after [300] seconds waiting for service container stability. Operation will roll back. Step that first updated the service container was 'add' at address '[("deployment" => "FineWebFailoverTestCase.war")]'
&amp#27;[0m&amp#27;[31m08:12:57,040 ERROR [org.jboss.as.server] (management-handler-thread - 1) WFLYSRV0022: Deploy of deployment "FineWebFailoverTestCase.war" was rolled back with no failure message
&amp#27;[0m&amp#27;[31m08:13:02,040 ERROR [org.jboss.as.controller.management-operation] (management-handler-thread - 1) WFLYCTL0190: Step handler org.jboss.as.server.deployment.DeploymentHandlerUtil$1@659fd643 for operation add at address [("deployment" => "FineWebFailoverTestCase.war")] failed -- java.util.concurrent.TimeoutException: java.util.concurrent.TimeoutException
  at org.jboss.as.controller.OperationContextImpl.waitForRemovals(OperationContextImpl.java:538)
  at org.jboss.as.controller.AbstractOperationContext$Step.handleResult(AbstractOperationContext.java:1566)
  at org.jboss.as.controller.AbstractOperationContext$Step.finalizeInternal(AbstractOperationContext.java:1520)
  at org.jboss.as.controller.AbstractOperationContext$Step.finalizeStep(AbstractOperationContext.java:1493)
  at org.jboss.as.controller.AbstractOperationContext$Step.access$400(AbstractOperationContext.java:1367)
  at org.jboss.as.controller.AbstractOperationContext.executeResultHandlerPhase(AbstractOperationContext.java:912)
  at org.jboss.as.controller.AbstractOperationContext.processStages(AbstractOperationContext.java:762)
  at org.jboss.as.controller.AbstractOperationContext.executeOperation(AbstractOperationContext.java:468)
  at org.jboss.as.controller.OperationContextImpl.executeOperation(OperationContextImpl.java:1428)
  at org.jboss.as.controller.ModelControllerImpl.internalExecute(ModelControllerImpl.java:448)
  at org.jboss.as.controller.ModelControllerImpl.lambda$executeForResponse$0(ModelControllerImpl.java:259)
  at org.wildfly.security.auth.server.SecurityIdentity.runAs(SecurityIdentity.java:304)
  at org.wildfly.security.auth.server.SecurityIdentity.runAs(SecurityIdentity.java:270)
  at org.jboss.as.controller.ModelControllerImpl.executeForResponse(ModelControllerImpl.java:259)
  at org.jboss.as.controller.ModelControllerImpl.executeOperation(ModelControllerImpl.java:253)
  at org.jboss.as.controller.ModelControllerImpl.execute(ModelControllerImpl.java:236)
  at org.jboss.as.controller.remote.ModelControllerClientOperationHandler$ExecuteRequestHandler.doExecute(ModelControllerClientOperationHandler.java:241)
  at org.jboss.as.controller.remote.ModelControllerClientOperationHandler$ExecuteRequestHandler.access$400(ModelControllerClientOperationHandler.java:139)
  at org.jboss.as.controller.remote.ModelControllerClientOperationHandler$ExecuteRequestHandler$1$1.run(ModelControllerClientOperationHandler.java:163)
  at org.jboss.as.controller.remote.ModelControllerClientOperationHandler$ExecuteRequestHandler$1$1.run(ModelControllerClientOperationHandler.java:159)
  at org.wildfly.security.auth.server.SecurityIdentity.runAs(SecurityIdentity.java:328)
  at org.wildfly.security.auth.server.SecurityIdentity.runAs(SecurityIdentity.java:285)
  at org.jboss.as.controller.AccessAuditContext.doAs(AccessAuditContext.java:254)
  at org.jboss.as.controller.AccessAuditContext.doAs(AccessAuditContext.java:225)
  at org.jboss.as.controller.remote.ModelControllerClientOperationHandler$ExecuteRequestHandler$1.execute(ModelControllerClientOperationHandler.java:159)
  at org.jboss.as.protocol.mgmt.ManagementRequestContextImpl$1.doExecute(ManagementRequestContextImpl.java:70)
  at org.jboss.as.protocol.mgmt.ManagementRequestContextImpl$AsyncTaskRunner.run(ManagementRequestContextImpl.java:160)
  at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
  at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1990)
  at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1486)
  at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1377)
  at java.lang.Thread.run(Thread.java:748)
  at org.jboss.threads.JBossThread.run(JBossThread.java:513)
```

**Review by rhusar — APPROVED** [review]
LGTM!

The failure in 'Linux - Elytron - JDK 8' is to be expected and is addressed by https://github.com/wildfly/wildfly/pull/14533 depending on this PR.

**Pull request merged**